File: Cpp/351.cpp
class Solution {
private:
    vector<vector<int>> skip;

public:
    Solution() {
        skip.assign(10, vector<int>(10, 0));
        skip[1][3] = skip[3][1] = 2;
        skip[1][7] = skip[7][1] = 4;
        skip[3][9] = skip[9][3] = 6;
        skip[7][9] = skip[9][7] = 8;
        skip[2][8] = skip[8][2] = skip[4][6] = skip[6][4] = skip[1][9]
            = skip[9][1] = skip[3][7] = skip[7][3] = 5;
    }

    int numberOfPatterns(int m, int n) {
        int res = 0;
        vector<bool> visit(10, false);
        for (int i = m; i <= n; ++i) {
            res += dfs(visit, 1, i - 1) * 4;
            res += dfs(visit, 2, i - 1) * 4;
            res += dfs(visit, 5, i - 1);
        }

        return res;
    }

    int dfs(vector<bool> visit, int start, int remaining) {
        if (remaining == 0) {
            return 1;
        }

        int res = 0;
        visit[start] = true;
        for (int i = 1; i <= 9; ++i) {
            if (visit[i]) {
                continue;
            }
            if (skip[start][i] == 0 || visit[skip[start][i]]) {
                res += dfs(visit, i, remaining - 1);
            }
        }

        return res;
    }
};

File: Python/351.py
class Solution(object):
    def __init__(self):
        self.skip = [[0]*10 for _ in range(10)]
        self.skip[1][3] = self.skip[3][1] = 2
        self.skip[1][7] = self.skip[7][1] = 4
        self.skip[3][9] = self.skip[9][3] = 6
        self.skip[7][9] = self.skip[9][7] = 8
        self.skip[1][9] = self.skip[9][1] = self.skip[3][7] = self.skip[7][3] = \
        self.skip[2][8] = self.skip[8][2] = self.skip[4][6] = self.skip[6][4] = 5
        
    
    def numberOfPatterns(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        visit = [False]*10
        def dfs(start, remaining, visit):
            if remaining == 0:
                return 1
            res = 0
            visit[start] = True
            for i in range(1, 10):
                if visit[i]:
                    continue
                if self.skip[start][i] == 0 or visit[self.skip[start][i]]:
                    res += dfs(i, remaining-1, visit)
            visit[start] = False
            return res
        
        
        res = 0
        for i in range(m, n+1):
            res += dfs(1, i-1, visit)*4
            res += dfs(2, i-1, visit)*4
            res += dfs(5, i-1, visit)
        return res
